### PR TITLE
Catch and report exceptions in guards like preconditions

### LINKED ISSUE #263: Matching of variant guards should catch exceptions
`(s/validate (s/conditional even? s/Int) "1")` 

throws an exception when it should behave the same as:

`(s/check (s/pred even?) "1")`

(returns an error like `(throws? (even? "1"))`)

### PR DESCRIPTION
**Pull request by w01fe**
Fixes #263

### CODE REVIEW

File: src/cljx/schema/spec/variant.cljx [hunk 1]
```diff
@@ -13,7 +13,17 @@
   (let [g (:guard o)
         c (spec/sub-checker o params)
         step (if g
-               (fn [x] (if (g x) (c x) (else x)))
+               (fn [x]
+                 (macros/try-catchall
+                  (if (g x)
```
Comment (loganlinn) [inline src/cljx/schema/spec/variant.cljx]: is there merit to tightening the try-catch to be just around `(g x)` and not around sub-checker/else?

Comment (w01fe) [inline src/cljx/schema/spec/variant.cljx]: Good call, will update.

Comment (w01fe) [inline src/cljx/schema/spec/variant.cljx]: (It's just annoying since now we need trinary, unless you see another way around it?)

**Comment by w01fe** [comment]
I guess an alternative is to just consider the guard false when it fails.  This will generally do the right thing, it just seems a bit gross and potentially more confusing in some cases.  Thoughts?

**Comment by loganlinn** [comment]
Agree it's generally right behavior, but error messages could be
misleading. They way you have it now LGTM

**Pull request merged**